babyProject/babyProject/Views/ChannelViewController.swift

```diff
@@ -0,0 +1,118 @@
+//
+//  ChannelViewController.swift
+//  babyProject
+//
+//  Created by Otto Linden on 24.4.2021.
+//
+
+import UIKit
+import MessageKit
+import InputBarAccessoryView
+import Combine
+import StompClientLib
+
+class ChannelViewController: MessagesViewController, MessagesDataSource, MessagesLayoutDelegate, MessagesDisplayDelegate {
+
+    private var cancellable: AnyCancellable?
+
+    let currentUser = Sender(senderId: "1", displayName: "iOS-User-\(UUID().uuidString)")
+    var _messages = [Message]()
+    var messages: [Message] {
+        get {
+            return _messages
+        } set {
+            _messages = newValue
+            messagesCollectionView.reloadData()
+        }
+    }
+    var channel: bpChannel?
+
+    var socketClient = StompClientLib()
+    lazy var url = NSURL(string: "https://api.turkuforge.fi/connect/websocket")!
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        messagesCollectionView.messagesDataSource = self
+        messagesCollectionView.messagesLayoutDelegate = self
+        messagesCollectionView.messagesDisplayDelegate = self
+        messageInputBar.delegate = self
+
+        socketClient.openSocketWithURLRequest(request: NSURLRequest(url: url as URL) , delegate: self)
+    }
+
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        navigationItem.largeTitleDisplayMode = .never
+    }
+
+    func currentSender() -> SenderType {
+        return currentUser
+    }
+
+    func messageForItem(at indexPath: IndexPath, in messagesCollectionView: MessagesCollectionView) -> MessageType {
+        return messages[indexPath.section]
+    }
+    
+    func numberOfSections(in messagesCollectionView: MessagesCollectionView) -> Int {
+        return messages.count
+    }
+
+    func send(message: String) {
+        if let url = channel?.links.linkSelf?.href {
+            var request = URLRequest(url: url)
+            request.httpMethod = "POST"
+            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
+            let data = [
+                "content": message,
+                "sender": currentSender().displayName
+            ]
+            let jsonData = try! JSONSerialization.data(withJSONObject: data, options: [])
+            URLSession.shared.uploadTask(with: request, from: jsonData) { data, response, error in
+            }.resume()
+        }
+    }
+
+    func received(message: RecivedMessage) {
+        let message = Message(sender: Sender(senderId: message.from == currentUser.displayName ? "1" : "2", displayName: message.from), messageId: "2", sentDate: message.date, kind: .text(message.message))
```

Comment:
Would it work if this was moved to be like this.
```suggestion
        let sender = Sender(
            senderId: message.from == currentUser.displayName ? "1" : "2", 
            displayName: message.from
        )
        let message = Message(
            sender: sender, 
            messageId: "2", 
            sentDate: message.date, 
            kind: .text(message.message)
        )
```

Not sure if you can do that in swift but if possible would be a bit easyer to read.

Also what is the condision about `message.from == currentUser.displayName ? "1" : "2"` Don't fully understand the logic there, also what does the `messageId` do and why is it hardcoded to "2"?

Reply:
We need to specify a `SenderID` to know who sent the message and so that it goes on the correct side of the screen (left side incoming messages, right side outgoing messages), but since we don't have them in our API yet I decided that the current user will have `SenderID 1` and everyone else 2. 

So when we receive the messages from the server I check if the sender of that message has the same name as we do and then sets the `SenderID` according to that.

This is of course not a good solution, but it will have to work for now as our API is not complete.

A `MessageID` is for future use (e.g reference to a message), but that is something that `MessageKit` requires us to provide. It can be anything, I should probably have documented that a little better and that it is only hardcoded for now.